### PR TITLE
INPUT: Add gamepad & mouse support

### PR DESCRIPTION
**Pull request by Fusioon**
Added some basic mouse and gamepad support, moved input handling from `Application` into `GLFWInputManager`, also added function for binding `KeyCodes` to string and then getting keystate with that. Should support multiple gamepads but that is something I can't test.

My next plan for the input system is to make something along the lines of UE4 input. 
Developer defines Action which contain key or axis bindings then in code they register callback function which will be called when given action occurs. These actions can be stored into ActionMap (basically a collection of Actions when they can all be enabled/disabled) which will allow for easy switching of controls if player enters a vehicle for example. 
I would like to hear what you think about this.

This PR will create some conflicts when we will want to merge #20.

### CODE REVIEW

**Comment by jamesaorson** [comment]
> Added some basic mouse and gamepad support, moved input handling from `Application` into `GLFWInputManager`, also added function for binding `KeyCodes` to string and then getting keystate with that. Should support multiple gamepads but that is something I can't test.
> 
> My next plan for the input system is to make something along the lines of UE4 input.
> Developer defines Action which contain key or axis bindings then in code they register callback function which will be called when given action occurs. These actions can be stored into ActionMap (basically a collection of Actions when they can all be enabled/disabled) which will allow for easy switching of controls if player enters a vehicle for example.
> I would like to hear what you think about this.
> 
> This PR will create some conflicts when we will want to merge #20.

Since this PR is much smaller, I suggest we merge this one before #20 and fix the conflicts there. Then we can add some input testing to the example components created in #20.

File: Engine/src/Application.bf [hunk 1]
```diff
@@ -11,6 +11,7 @@ namespace SteelEngine
 
 		private Window _window ~ delete _;
 		private Window.EventCallback _eventCallback = new => OnEvent ~ delete _;
+		private GLFWInputSystem _inputSystem = new GLFWInputSystem() ~ delete _;
```
Comment (jamesaorson) [inline Engine/src/Application.bf]: With the introduction of Systems in #20, can we rename this to `GLFWInputManager` or something of that sort?

File: Engine/src/Input/AxisCode.bf [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+		MouseX = 0,
+		MouseY,
+		MouseScroll,
+
+		GamepadLeftStickX,
+		GamepadLeftStickY,
+		GamepadRightStickX,
+		GamepadRightStickY,
+		GamepadLeftTrigger,
+		GamepadRightTrigger,
```
Comment (jamesaorson) [inline Engine/src/Input/AxisCode.bf]: Is the Gamepad prefix necessary here? Be nice to have this somewhat shortened if we never anticipate a conflict in this enum.

File: Engine/src/Input/AxisCode.bf [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	}
+
+	public enum GamepadAxisCode : AxisCode
+	{
+		LeftStickX = (.)AxisCode.GamepadLeftStickX,
+		LeftStickY = (.)AxisCode.GamepadLeftStickY,
+		LeftTrigger = (.)AxisCode.GamepadLeftTrigger,
+		RightStickX = (.)AxisCode.GamepadRightStickX,
+		RightStickY = (.)AxisCode.GamepadRightStickY,
+		RightTrigger = (.)AxisCode.GamepadRightTrigger,
```
Comment (jamesaorson) [inline Engine/src/Input/AxisCode.bf]: Why is this necessary if we have AxisCode already? Is this just to be more explicit about intent?

File: Engine/src/Input/GLFWInputSystem.bf [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+using SteelEngine.Window;
+using glfw_beef;
+
+namespace SteelEngine.Input
+{
+	public class GLFWInputSystem
+	{
+		HashSet<int> _connectedGamepads = new HashSet<int>() ~ delete _;
+
+		public Result<void> Initialize()
```
Comment (jamesaorson) [inline Engine/src/Input/GLFWInputSystem.bf]: Why have a `Result<void>` return type if the only return value is `.Ok`?

File: Engine/src/Input/AxisCode.bf (hunk 3 above)
Comment (jamesaorson) [reply Engine/src/Input/AxisCode.bf]: Okay. Looking later on in the file, I see this is being used for indexing into an array. Can we be explicit about this use case as a comment on this enum?

File: Engine/src/Input/GLFWKeyMapper.bf [hunk 5]
```diff
@@ -163,8 +179,31 @@ namespace SteelEngine.Input
...
+				case .DPadRight: return .GamepadRight;
+				case .DPadUp: return .GamepadUp;
+				case .DPadDown: return .GamepadDown;
+				case .LeftThumb: return .GamepadLeftStick;
+				case .RightThumb: return .GamepadRightStick;
+				case .Start: return .GamepadStart;
+				case .Guide: return .GamepadHome;
+				case .Back: return .GamepadSelect;
+				case .LeftBumper: return .GamepadL1;
+				case .RightBumper: return .GamepadR1;
```
Comment (jamesaorson) [inline Engine/src/Input/GLFWKeyMapper.bf]: If we are going to go with `A, B, X, Y`, we should stay consistent with XBOX controller naming standards and keep these as `Back, LeftBumper, RightBumper`.

File: Engine/src/Input/GamepadInfo.bf [hunk 6]
```diff
@@ -0,0 +1,91 @@
+using System;
+namespace SteelEngine.Input
+{
+	public class GamepadInfo
+	{
+		const let KeyCodeStart = KeyCode.GamepadA;
+		const let KeyCodeEnd = KeyCode.GamepadRightStick + 1;
+		const let AxisCodeStart = GamepadAxisCode.LeftStickX;
+		const let AxisCodeEnd = GamepadAxisCode.RightTrigger + 1;
```
Comment (jamesaorson) [inline Engine/src/Input/GamepadInfo.bf]: Since these are `const`, they should also be uppercase like `NUMBER_OF_KEYS`.

File: Engine/src/Input/GamepadInfo.bf [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+		const int NUMBER_OF_KEYS = int(KeyCodeEnd-KeyCodeStart);
+		const int NUMBER_OF_AXIS = int(AxisCodeEnd-AxisCodeStart);
```
Comment (jamesaorson) [inline Engine/src/Input/GamepadInfo.bf]: Plural for `axis` is `axes`

File: Engine/src/Input/Input.bf [hunk 8]
```diff
@@ -10,40 +11,121 @@ namespace SteelEngine.Input
...
 			_accumulatedEvents = default;
+			_accumulatedAxisValues = default;
 		}
 
 		public static void ResetInput()
 		{
 			_accumulatedEvents = default;
 			_lastUpdateState = default;
+			_axisValues = default;
+			for (let pad in _gamepads.Values) pad.ResetInput();
```
Comment (jamesaorson) [inline Engine/src/Input/Input.bf]: To be consistent with other one-liner for loops, move `pad.ResetInput()` to the next line.

**Review by jamesaorson — CHANGES_REQUESTED** [review]

File: Engine/src/Input/GLFWInputSystem.bf (hunk 4 above)
Comment (Fusioon) [reply Engine/src/Input/GLFWInputSystem.bf]: I feel like Initialize methods should be able to fail, even if it's currently not used. It may save us some trouble later.

File: Engine/src/Input/GLFWKeyMapper.bf (hunk 5 above)
Comment (Fusioon) [reply Engine/src/Input/GLFWKeyMapper.bf]: I'm still considering renaming the A, B, X, Y buttons to something more generic like South, West, East, North. But yes it should be more consistent.

File: Engine/src/Input/AxisCode.bf (hunk 3 above)
Comment (Fusioon) [reply Engine/src/Input/AxisCode.bf]: Yeah, that enum is not really necesary. Could add check to the functions that index the array if the value is in range and if not it generates error and returns default.

File: Engine/src/Input/GLFWInputSystem.bf (hunk 4 above)
Comment (jamesaorson) [reply Engine/src/Input/GLFWInputSystem.bf]: I will also add this to the initialize methods of the ECS

File: Engine/src/Input/GLFWKeyMapper.bf (hunk 5 above)
Comment (jamesaorson) [reply Engine/src/Input/GLFWKeyMapper.bf]: People may not know their compass directions. I would advocate for the XBOX controller reference, or `Up, Down, Left, Right`, and change DPad to `DPadUp, DPadDown, DPadLeft, DPadRight`

**Review by RogueMacro — APPROVED** [review]
